tensorflow/core/kernels/parse_tensor_test.cc

```diff
@@ -0,0 +1,118 @@
+/* Copyright 2017 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include <memory>
+#include <vector>
+#include <string>
+
+#include "tensorflow/core/common_runtime/device.h"
+#include "tensorflow/core/common_runtime/device_factory.h"
+#include "tensorflow/core/framework/allocator.h"
+#include "tensorflow/core/framework/fake_input.h"
+#include "tensorflow/core/framework/node_def_builder.h"
+#include "tensorflow/core/framework/op_kernel.h"
+#include "tensorflow/core/framework/tensor.h"
+#include "tensorflow/core/framework/types.h"
+#include "tensorflow/core/framework/types.pb.h"
+#include "tensorflow/core/kernels/ops_testutil.h"
+#include "tensorflow/core/kernels/ops_util.h"
+#include "tensorflow/core/framework/register_types.h"
+
+namespace tensorflow {
+namespace {
+
+class SerializeTensorOpTest : public OpsTestBase {
+ protected:
+  template <typename T>
+  void MakeOp() {
+    TF_ASSERT_OK(
+        NodeDefBuilder("myop", "SerializeTensor")
+            .Input(FakeInput(DataTypeToEnum<T>::value))
+            .Finalize(node_def()));
+    TF_ASSERT_OK(InitOp());
+  }
+};
+
+#define REGISTER_TEST(T, Shape, Mapping)                                 \
```

Comment:
Thanks for adding the test!

I think it would be preferable not to use a huge macro for the entire test though:

- It's harder to know the name of the test to run because the full name is hidden behind a macro expansion
- There are two levels of macros here, which makes it harder to read.

Instead, can you do the following?

* Define each TEST_F manually with a name that is fixed
* define helper functions templated by T that does things like MakeOp, AddInput, NodeDefBuilder, etc, but keep parts that are not templated by T outside.
* define helper functions that do all of the other boiler plate not templated by T
* have each test call those helper functions with the appropriate fixed type.